examples/image-search/build.cpp

```diff
@@ -0,0 +1,155 @@
+}
+
+// returns success
+bool my_app_params_parse(int argc, char **argv, my_app_params &params) {
+	bool invalid_param = false;
+	for (int i = 1; i < argc; i++) {
+
+		std::string arg = argv[i];
+
+		if (arg == "-m" || arg == "--model") {
+			if (++i >= argc) {
+				invalid_param = true;
+				break;
+			}
+			params.model = argv[i];
+		} else if (arg == "-t" || arg == "--threads") {
+			if (++i >= argc) {
+				invalid_param = true;
+				break;
+			}
+			params.n_threads = std::stoi(argv[i]);
+		} else if (arg == "-h" || arg == "--help") {
+			my_print_help(argc, argv, params);
+			exit(0);
+		} else if (arg.starts_with('-')) {
+			if (i != 0) {
+				printf("%s: unrecognized argument: %s\n", __func__, arg.c_str());
+				return false;
+			}
+		} else {
+			// assume image directory
+			params.image_directories.push_back(argv[i]);
+		}
+	}
+
+	return !(invalid_param || params.image_directories.empty());
+}
+
+bool is_image_file_extension(std::string_view ext) {
+	if (ext == ".jpg") return true;
+	if (ext == ".JPG") return true;
+
+	if (ext == ".jpeg") return true;
+	if (ext == ".JPEG") return true;
+
+	if (ext == ".gif") return true;
+	if (ext == ".GIF") return true;
+
+	if (ext == ".png") return true;
+	if (ext == ".PNG") return true;
+
+	// TODO: add all the supported files for std_image
+
+	return false;
+}
+
+int main(int argc, char** argv) {
+	my_app_params params;
+	if (!my_app_params_parse(argc, argv, params)) {
+		my_print_help(argc, argv, params);
+		return 1;
+	}
+
+	auto clip_ctx = clip_model_load(params.model.c_str());
+	if (!clip_ctx) {
+		printf("%s: Unable  to load model from %s\n", __func__, params.model.c_str());
+		return 1;
+	}
+
+	std::vector<std::string> image_file_index;
+	unum::usearch::index_gt<unum::usearch::cos_gt<float>> embd_index;
+
+	const size_t vec_dim = clip_ctx->vision_model.hparams.projection_dim;
+
+	size_t label = 0;
+
+	// search for images in path and create embedding
+	for (const auto& base_dir : params.image_directories) {
+		fprintf(stdout, "%s: starting base dir scan of '%s'\n", __func__, base_dir.c_str());
+
+		for (auto const& dir_entry : std::filesystem::recursive_directory_iterator(base_dir)) {
+			if (!dir_entry.is_regular_file()) {
+				continue;
+			}
+
+			// check for image file
+			const auto& ext = dir_entry.path().extension();
+			if (ext.empty()) {
+				continue;
+			}
+			if (!is_image_file_extension(ext.c_str())) {
+				continue;
+			}
+
+			std::string img_path {dir_entry.path()};
+			fprintf(stdout, "%s: found image file '%s'\n", __func__, img_path.c_str());
+
+			clip_image_u8 img0;
+			clip_image_f32 img_res;
+			if (!clip_image_load_from_file(img_path, img0)) {
+				fprintf(stderr, "%s: failed to load image from '%s'\n", __func__, img_path.c_str());
+				continue;
+			}
+
+			clip_image_preprocess(&img0, &img_res);
+
+			std::vector<float> vec(vec_dim);
+			if (!clip_image_encode(clip_ctx, params.n_threads, img_res, vec.data())) {
```

Comment:
Why not using a float array directly? We are working with fixed lengths at the end.

Reply:
are we? I thought it depends on the specific model used.
but yea, i guess we could reuse the same vector, since it does not change during execution.

Reply:
Yeah, it's more like a coding style issue. both this repo and the USearch repo use float arrays instead of `std::vector<float>` to represent embedding vectors.

Reply:
> float arrays

but the size is not known at compile time. runtime size declaration is an extension, and not proper c++

Reply:
I just meant, for example, copying from `examples/main.cpp`

```C++
    float img_vec[vec_dim];
    if (!clip_image_encode(ctx, params.n_threads, img_res, img_vec))
    {
        return 1;
    }
```

But ok, it was just a nit, and not a strong opinion :).